File: nic/upgrade_manager/export/upggosdk/upgrade_ctx_api.go
package upggosdk

import (
	"errors"

	"github.com/pensando/sw/nic/upgrade_manager/export/upggosdk/nic/upgrade_manager/upgrade"
)

//UpgState  of the state machine
type UpgState int

const (
	//COMPATCHECK state
	COMPATCHECK UpgState = 0
	//PRESWITCHROOT state
	PRESWITCHROOT UpgState = 1
	//POSTSWITCHROOT state
	POSTSWITCHROOT UpgState = 1
)

//UpgCtxGetPreUpgTableVersion is used to get the version of a component identified by meta
// in pre-upgrade image
func UpgCtxGetPreUpgTableVersion(ctx *UpgCtx, meta CompVersion) (string, error) {
	switch meta {
	case NICMGR:
		return ctx.PreUpgMeta.NicmgrVersion, nil
	case KERNEL:
		return ctx.PreUpgMeta.KernelVersion, nil
	case PCIE:
		return ctx.PreUpgMeta.PcieVersion, nil
	}
	return "", errors.New("Version not found")
}

//UpgCtxGetPostUpgTableVersion is used to get the version of a component identified by meta
// in post-upgrade image
func UpgCtxGetPostUpgTableVersion(ctx *UpgCtx, meta CompVersion) (string, error) {
	switch meta {
	case NICMGR:
		return ctx.PostUpgMeta.NicmgrVersion, nil
	case KERNEL:
		return ctx.PostUpgMeta.KernelVersion, nil
	case PCIE:
		return ctx.PostUpgMeta.PcieVersion, nil
	}
	return "", errors.New("Version not found")
}

//UpgCtxIsUpgTypeNonDisruptive used to know if it is non-disruptive upgrade
func UpgCtxIsUpgTypeNonDisruptive(ctx *UpgCtx) bool {
	return ctx.upgType == upgrade.UpgType_UpgTypeNonDisruptive
}

//UpgCtxIsUpgTypeDisruptive used to know if it is disruptive upgrade
func UpgCtxIsUpgTypeDisruptive(ctx *UpgCtx) bool {
	return ctx.upgType == upgrade.UpgType_UpgTypeDisruptive
}

//UpgCtxGetUpgState used to get previous upgrade state
func UpgCtxGetUpgState(ctx *UpgCtx) UpgState {
	switch ctx.prevExecState {
	case upgrade.UpgReqStateType_UpgStateCompatCheck:
		return COMPATCHECK
	case upgrade.UpgReqStateType_UpgStateLinkDown:
		return PRESWITCHROOT
	case upgrade.UpgReqStateType_UpgStateHostDown:
		return PRESWITCHROOT
	case upgrade.UpgReqStateType_UpgStatePostHostDown:
		return PRESWITCHROOT
	case upgrade.UpgReqStateType_UpgStateSaveState:
		return PRESWITCHROOT

	case upgrade.UpgReqStateType_UpgStatePostRestart:
		return POSTSWITCHROOT
	case upgrade.UpgReqStateType_UpgStateHostUp:
		return POSTSWITCHROOT
	case upgrade.UpgReqStateType_UpgStateLinkUp:
		return POSTSWITCHROOT
	case upgrade.UpgReqStateType_UpgStatePostLinkUp:
		return POSTSWITCHROOT
	}
	return COMPATCHECK
}

File: nic/upgrade_manager/export/upgcsdk/upgrade_ctx_api.cc
// {C} Copyright 2018 Pensando Systems Inc. All rights reserved.

#include <stdio.h>
#include <iostream>

#include "upgrade_ctx_api.hpp"
#include "nic/upgrade_manager/utils/upgrade_log.hpp"

namespace upgrade {

using namespace std;

delphi::error UpgCtxApi::UpgCtxGetUpgTableVersion (ImageInfo& imgInfo, UpgMeta meta, string &version) {
    switch (meta) {
        case NICMGRVER:
            version = imgInfo.nicmgrVersion;
            return delphi::error::OK();
        case KERNELVER:
            version = imgInfo.kernelVersion;
            return delphi::error::OK();
    }
    return delphi::error("Meta not found");
}

delphi::error UpgCtxApi::UpgCtxGetPreUpgTableVersion (UpgCtx &ctx, UpgMeta meta, string &version) {
    return UpgCtxGetUpgTableVersion(ctx.preUpgMeta, meta, version);
}

delphi::error UpgCtxApi::UpgCtxGetPostUpgTableVersion (UpgCtx &ctx, UpgMeta meta, string &version) {
    return UpgCtxGetUpgTableVersion(ctx.postUpgMeta, meta, version);
}

bool UpgCtxApi::UpgCtxIsUpgTypeDisruptive(UpgCtx &ctx) {
    return ctx.upgType == UpgTypeDisruptive;
}

bool UpgCtxApi::UpgCtxIsUpgTypeNonDisruptive(UpgCtx &ctx) {
    return ctx.upgType == UpgTypeNonDisruptive;
}

}
